### ISSUE TITLE
FormBuilder unexpected values with presenters

### ISSUE THREAD

**Issue by dikond** [hanami/helpers#91]
I am passing a presenter instance to the `form_for` helper, which decorates some of the methods of my model (I use plain Sequel, without repositories), and I want the form builder to call that decorated method.

When the form initialized, [it calls](https://github.com/hanami/helpers/blob/master/lib/hanami/helpers/form_helper/values.rb#L21) `.symbolize!` on the values, which [recursively calls itself](https://github.com/hanami/utils/blob/master/lib/hanami/utils/hash.rb#L68) on the values which respond to `.to_hash`.

While this behaviour may be handy, it produces unexpected results when we have presenter object which responds to `to_hash` in the `values` options.

My solution to the problem was decorating `respond_to? :to_hash` to return `false` for the instance of my presenter. While it works, I don't think it is explicit behaviour.

Let me know if you would need any help with this.

**Comment by myabc** [comment]
@dikond what do you mean by "unexpected results"?

**Comment by dikond** [comment]
Imagine I have a model object that responds to following methods:
```ruby
party.invited_guests
#=> ["James", "Anna", "Vitalik"]
party.address
# "Stanislavskogo, 15"
```

And I have a presenter object that decorates `invited_guests` method:
```ruby
party_presenter.invited_guests
#=> "Anna, James, Vitalik"
```
Its class includes `Hanami::Presenter`


In the view I have a form to edit this:
```ruby
def form
  form_for :party, routes.party_path(party.id), values: { party: party_presenter }, method: :patch do
    text_field :address
    text_area :invited_guests
  end
end
```

In this case, the value of text_area would be `["James", "Anna", "Vitalik"]`, while I expect it to be `"Anna, James, Vitalik"`.

I expected this after reading a guide about [form helper](http://hanamirb.org/guides/helpers/forms/#automatic-values). There is no remark that object passed to `values` option will be converted to hash, which may not always be what we want, I guess...

**Comment by jodosha** [comment]
@dikond Can you please check if [this PR](https://github.com/hanami/helpers/pull/87) by @myabc fixes the problem for you? Thanks. 😄

**Comment by myabc** [comment]
@dikond if I understand correctly, your test case looks something like this:

```ruby
require 'hanami-view'
require 'hanami/utils/hash'
require 'pp'

class Party
  attr_accessor :invited_guests
  attr_accessor :address

  def to_hash
    { invited_guests: invited_guests, address: address }
  end
end

class PartyPresenter
  include Hanami::Presenter
  
  def invited_guests
    super.join(', ')
  end
end

party = Party.new
party.invited_guests  = ["James", "Anna", "Vitalik"]
party.address         = "Stanislavskogo, 15"

party_presenter = PartyPresenter.new(party)

p party_presenter.invited_guests
pp Hanami::Utils::Hash.new(party_presenter)
```

A easy workaround would be for your `PartyPresenter` to define `to_hash`:
```ruby
  def to_hash
    super.merge(invited_guests: invited_guests)
  end
```

I'll leave defining what the right behaviour should be up to @jodosha though!

**Comment by myabc** [comment]
To summarise, you will get expected behaviour if:
- both model and presenter implement `#to_hash`
- neither model nor presenter implement `#to_hash` (e.g. you add `undef_method :to_hash` to your `Party` class, if subclassing `Hanami::Entity`).

I'd suggest that this is actually an issue with `Hanami::View` - `Hanami::Presenter` should make this caveat clear (or provide a solution).

**Comment by dikond** [comment]
@jodosha sure, I've checked that PR before opening my issue. The thing is that an instance of presenter gets converted to `Utils::Hash`, thus, passing [this line](https://github.com/hanami/helpers/pull/87/files#diff-c423ddc54c30b5b2e624939494dc38e7R52).

@myabc yeah, the case is like that. My conclusions was almost the same. We can also prepare our data explicitly, e.g. converting it to hash and passing it with `values` in the `form_for`. But I do not like the idea of `undef`, because in some cases we may need `to_hash` on presenter object, so I did something like this:

```ruby
# helper in the view
def decorate(model, convertable: true)
  instance = presenter.new(model)
  instance.extend(Core::Ext::UnresponsiveToHash) unless convertable
  instance
end

# somewhere in the code
module Core
  module Ext
    module UnresponsiveToHash
      def respond_to?(name, *opts)
        return false if name.intern == :to_hash
        super
      end
    end
  end
end

# and in the view
def area_presenter
  decorate area, convertable: false
end

def form
  form_for :area, url, values: { area: area_presenter} do #...
end
```

**Comment by myabc** [comment]
@dikond I think a simpler, generic fix would be to add a black-list of methods that `Hanami::Presenter` should not delegate. e.g.:

```diff
+++ b/lib/hanami/view/escape.rb
@@ -146,7 +146,7 @@ module Hanami
         # @api private
         # @since 0.7.0
         def method_missing(m, *args, &blk)
-          if @object.respond_to?(m)
+          if !exclude_delegation_methods.include?(m) && @object.respond_to?(m)
             ::Hanami::View::Escape.html(@object.__send__(m, *args, &blk))
           else
             super
@@ -158,7 +158,12 @@ module Hanami
         # @api private
         # @since 0.3.0
         def respond_to_missing?(m, include_private = false)
-          @object.respond_to?(m, include_private)
+          !exclude_delegation_methods.include?(m) &&
+            @object.respond_to?(m, include_private)
+        end
+
+        def exclude_delegation_methods
+          %i(to_hash)
         end
       end
 ```

I'd wait from feedback from @jodosha before submitting such a patch though.

**Comment by kaikuchn** [comment]
I'm experiencing a similiar problem, although no presenter is involved, instead i have a `Virtus.value_object`.
```ruby
form_for :alert, routes.alert_path(id: alert.id), method: :patch, values: { alert: alert }
```
Where alert is said value object.
Since [Utils::Hash#symbolize!](https://github.com/hanami/utils/blob/master/lib/hanami/utils/hash.rb#L68) is called [here](https://github.com/hanami/helpers/blob/e19ad89a81bfb7a09c268ed0c0bb3061f0d2d212/lib/hanami/helpers/form_helper/values.rb#L21) my alert poro is converted into a Hash and no longer responds to the methods it is supposed to. Thus the automatic filling of the form fails.

Maybe Utils::Hash#symbolize! shouldn't aggressively convert non-hashes to hashes when deeply symbolizing keys? Although I'm super new to Hanami, so I have no clue if this interferes with nested forms.

**Comment by kaikuchn** [comment]
Ah, I guess this will be resolve soon, see [this issue](https://github.com/hanami/utils/issues/178).

**Comment by jodosha** [comment]
@kaikuchn Hey, we released v0.9.2 on yesterday which ships with that fix. Can you please update and check if you still have the same problem? Thanks!

**Comment by kaikuchn** [comment]
Works with 0.9.2 as expected.

**Comment by mereghost** [comment]
Thanks for letting us know, @kaikuchn . I'll close the issue them.

**Comment by jodosha** [comment]
@kaikuchn Thanks for getting back. 👍

**Comment by dikond** [comment]
@mereghost @jodosha I apologise, but this was about presenters. I've test it with `hanami-helpers 0.5.1` & `hanami-utils 0.9.2` and can say that it still doesn't work.

**Comment by kaikuchn** [comment]
@dikond apologies for hijacking your issue!! I didn't mean to do that. It sounded like the reason for the weird behaviour was the same. 😖 

I mean, it shouldn't matter whether it's a presenter or some other object that responds to `to_hash`. And [Utils::Hash#symbolize!](https://github.com/hanami/utils/blob/master/lib/hanami/utils/hash.rb#L68) now only converts the keys and doesn't touch the values. I.e., in your example the values hash `{ party: party_presenter}` shouldn't convert the presenter at all.

What would explain the behaviour is when the values hash passed to the form-helper already is a `Hanami::Utils::Hash`. For `Hanami::Utils::Hash#initialize` calls `to_hash` on the passed argument and `Hanami::Utils::Hash#to_hash` calls [to_hash](https://github.com/hanami/utils/blob/master/lib/hanami/utils/hash.rb#L248) on its values. (Btw. that's strange behaviour, why does it do that?) Otherwise I can't explain why the value would get converted given your example. But maybe my reasoning is off or I am missing something. Could you maybe provide a simple failing test that demonstrates the behaviour?

**Comment by dikond** [comment]
I'll try to make a sample app or a failing test to demonstrate that, but won't be able to do that for the next 4-5 days.

**Comment by mereghost** [comment]
@dikond Hope to find you well. Any news on that?

**Comment by dikond** [comment]
@mereghost hi. This issue completely dropped out of my head. I'll provide the failing spec till the Monday.

**Comment by dikond** [comment]
@mereghost update:

I've checked the source of `hanami-utils 0.9.2` and... this version does not contain changes of https://github.com/hanami/utils/pull/162.

So I've tried to point the gem source to the github master. However, It requires to update almost every hanami gem to '1.0.0.beta1', and there is a funny issue:

```
hanami was resolved to 1.0.0.beta1, which depends on
      hanami-router (~> 1.0.0.beta1) was resolved to 1.0.0.beta1, which depends on
        rack (~> 2.0)
```
```
hanami-controller was resolved to 1.0.0.beta1, which depends on
      rack (>= 1.6.2, ~> 1.6)
```

`rack` version can't be resolved...

Anyway, I've made the same update as https://github.com/hanami/utils/pull/162 to the source code of my local hanami-utils gem. I confirm that with this changes my issues is resolved.
